Add support for loading an extension resources as a zip file.

xeenon opened the pull request:
#### df17cf322962aa8dbddc710428fb2334c13139e9
<pre>
Add support for loading an extension resources as a zip file.
<a href="https://webkit.org/b/281320">https://webkit.org/b/281320</a>
<a href="https://rdar.apple.com/136784811">rdar://136784811</a>

Reviewed by Brian Weinstein.

Support loading ZIP and CRX files directly with a WKWebExtension.

* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::extractTemporaryZipArchive): Added.
* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::createTemporaryZipArchive): Cleaned up.
(WTF::FileSystemImpl::extractTemporaryZipArchive): Added.
* Source/WebCore/en.lproj/Localizable.strings: Updated.
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.h:
(NS_ERROR_ENUM): Added WKWebExtensionErrorInvalidArchive.
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(+[WKWebExtension extensionWithResourceBaseURL:completionHandler:]): Remove directory assert.
(-[WKWebExtension _initWithAppExtensionBundle:resourceBaseURL:error:]): Ditto.
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::convertChromeExtensionToTemporaryZipFile): Added.
(WebKit::processFileAndExtractZipArchive): Added.
(WebKit::WebExtension::WebExtension): Use processFileAndExtractZipArchive when it is a directory.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::toAPI): Added InvalidArchive.
(WebKit::WebExtension::createError): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Tools/TestWebKitAPI/Tests/WebKit/web-extension.crx: Added.
* Tools/TestWebKitAPI/Tests/WebKit/web-extension.zip: Added.
* Tools/TestWebKitAPI/Tests/WebKit/web-extension/: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST(WKWebExtension, LoadFromDirectory)): Added.
(TestWebKitAPI::TEST(WKWebExtension, LoadFromZipArchive)): Added.
(TestWebKitAPI::TEST(WKWebExtension, LoadFromChromeExtensionArchive)): Added.
(TestWebKitAPI::TEST(WKWebExtension, LoadFromMacAppExtensionBundle)): Added.
(TestWebKitAPI::TEST(WKWebExtension, LoadFromiOSAppExtensionBundle)): Added.
(TestWebKitAPI::TEST(WKWebExtension, LoadWithBadURL)): Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/web-extension-ios.appex/: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/web-extension-mac.appex/: Added.

Canonical link: <a href="https://commits.webkit.org/285511@main">https://commits.webkit.org/285511@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3135b913a26eda075a7cd2a9e8ca5a0e53378dcb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72894 "56 style errors") | [❌ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52319 "Failed to checkout and rebase branch from PR 35450") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25694 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/77092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/24128 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/75009 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/60124 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23944 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/77092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/24128 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75961 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/47300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/62741 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/77092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/43947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/20209 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/22457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/66027 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/65802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/20565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/78765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/72150 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/17140 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/19711 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/78765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/86/builds/17188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/62747 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/78765 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/7009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93929 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11204 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/48117 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/2904 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/20665 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/49184 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50479 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/48929 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -106,27 +109,134 @@
+
+    static std::array<uint8_t, 4> expectedSignature = { 'C', 'r', '2', '4' };
+
+    // Verify Chrome extension magic signature.
```
Super nit: This comment could be above the `expectedSignature` to indicate that those bits are the `Chrome extension magic signature`.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -106,27 +109,134 @@
+
+            // Search for the ZIP file magic signature in the buffer.
+            for (ssize_t i = 0; i < bytesRead - 3; ++i) {
+                if (buffer[i] == 'P' && buffer[i + 1] == 'K' && buffer[i + 2] == 0x03 && buffer[i + 3] == 0x04) {
```
So are there two magic signatures in the file? This one and the `{ 'C', 'r', '2', '4' };`? I think this maybe could use a high level comment explaining what you are looking for.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
```diff
@@ -106,27 +109,134 @@
+
+static String processFileAndExtractZipArchive(const String& path)
+{
+    auto temporaryZipFilePath = convertChromeExtensionToTemporaryZipFile(path);
```
I think this would be more clear with an early return if `temporaryZipFilePath.isNull()`.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
```diff
@@ -1938,6 +1938,137 @@
+    [manager.get().defaultTab.webView loadRequest:urlRequest];
+
+    [manager run];
+}
```
Should this be PLATFORM(MAC)?

xeenon replied:
It passes on iOS, so I will just keep it as-is.

b-weinstein commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
```diff
@@ -1938,6 +1938,137 @@
+    [manager.get().defaultTab.webView loadRequest:urlRequest];
+
+    [manager run];
+}
```
And should this be PLATFORM(IOS)? I guess we want to be able to make sure to load iOS bundles on Mac.

xeenon replied:
We do need to make sure this works on Mac.

xeenon committed:
Add support for loading an extension resources as a zip file.
https://webkit.org/b/281320
rdar://136784811

Reviewed by Brian Weinstein.

Support loading ZIP and CRX files directly with a WKWebExtension.

* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::extractTemporaryZipArchive): Added.
* Source/WTF/wtf/FileSystem.h:
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::createTemporaryZipArchive): Cleaned up.
(WTF::FileSystemImpl::extractTemporaryZipArchive): Added.
* Source/WebCore/en.lproj/Localizable.strings: Updated.
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.h:
(NS_ERROR_ENUM): Added WKWebExtensionErrorInvalidArchive.
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtension.mm:
(+[WKWebExtension extensionWithResourceBaseURL:completionHandler:]): Remove directory assert.
(-[WKWebExtension _initWithAppExtensionBundle:resourceBaseURL:error:]): Ditto.
* Source/WebKit/UIProcess/API/Cocoa/WKWebExtensionPrivate.h:
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::convertChromeExtensionToTemporaryZipFile): Added.
(WebKit::processFileAndExtractZipArchive): Added.
(WebKit::WebExtension::WebExtension): Use processFileAndExtractZipArchive when it is a directory.
* Source/WebKit/UIProcess/Extensions/WebExtension.cpp:
(WebKit::toAPI): Added InvalidArchive.
(WebKit::WebExtension::createError): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
* Tools/TestWebKitAPI/Tests/WebKit/web-extension.crx: Added.
* Tools/TestWebKitAPI/Tests/WebKit/web-extension.zip: Added.
* Tools/TestWebKitAPI/Tests/WebKit/web-extension/: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtension.mm:
(TestWebKitAPI::TEST(WKWebExtension, LoadFromDirectory)): Added.
(TestWebKitAPI::TEST(WKWebExtension, LoadFromZipArchive)): Added.
(TestWebKitAPI::TEST(WKWebExtension, LoadFromChromeExtensionArchive)): Added.
(TestWebKitAPI::TEST(WKWebExtension, LoadFromMacAppExtensionBundle)): Added.
(TestWebKitAPI::TEST(WKWebExtension, LoadFromiOSAppExtensionBundle)): Added.
(TestWebKitAPI::TEST(WKWebExtension, LoadWithBadURL)): Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/web-extension-ios.appex/: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/web-extension-mac.appex/: Added.

Canonical link: https://commits.webkit.org/285511@main
